runtime/src/main/java/org/corfudb/runtime/collections/SMRMultiIndex.java

```diff
@@ -0,0 +1,362 @@
+     */
+    @FunctionalInterface
+    public interface ProjectionFunction<K, V, P>{
+        P generateProjection(K k, V v);
+    }
+
+    /**
+     * Data holder for User to be able to specify the index creation.
+     * It takes in a list of IndexFunction to index the main
+     * Map in different ways.
+     *
+     * @param <K> The type of the key.
+     * @param <V> The type of the value.
+     * @param <I> The type of the index key.
+     * @param <P> The type of projected value for the index
+     */
+    @Data
+    @EqualsAndHashCode(of = {"name"})
+    @AllArgsConstructor
+    public static class IndexSpecification<K, V, I,P>{
+        String name;
+        private List<IndexFunction<K, V, I>> indexFunctions;
+        private ProjectionFunction<K,V,P> projectionFunction;
+    }
+
+    /**
+     * To Specify multiple named indexes for the Main Map.
+     */
+    final Map<String, IndexSpecification> indexSpecifications;
+
+    /** The underlying HashMap. */
+    final Map<K, V> mainMap;
+
+
+    /**
+     * The Main Indexing Data Structure.
+     * Each named index maintains a <Code>Map</Code> of indexKeys that in turn
+     * point to a <Code>Map</Code> of keys and projection of values from the main <Code>Map</Code>
+     * datastructure. This multi-level <Code>Map</Code> provides flexibility for the current use case.
+     *
+     *  ________________________________________________________________________________________________________
+     * | <String>IndexName</String> -->  _______________________________________________________________________|
+     * |                                 | <I>IndexKey1</I> --> ________________________________________________|
+     * |                                 |                     |<K>PrimaryKey1</K> --><V>Projection(Value1)</V> |
+     * |                                 |                     |<K>PrimaryKey2</K> --><V>Projection(Value2)</V> |
+     * |                                 |                     |________________________________________________|
+     * |                                 |                     |________________________________________________|
+     * |                                 | <I>IndexKey2</I> -->|<K>PrimaryKey1</K> --><V>Projection(Value1)</V> |
+     * |                                 |                     |<K>PrimaryKey3</K> --><V>Projection(Value3)</V> |
+     * |_________________________________|_____________________|________________________________________________|
+     *
+     *
+     */
+    /** Indexes for the main map.*/
+    final Map<String, Map<I,Map<K,P>>> indexMaps;
+
+    /** Create a new in-memory multi-index given
+     * a set of indexing functions. */
+    public SMRMultiIndex(List<IndexSpecification> indexSpecifications) {
+
+        this.mainMap = new HashMap<>();
+        this.indexSpecifications = new HashMap<>();
+        indexMaps = new HashMap<String, Map<I,Map<K,P>>>();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace the type specification in this constructor call with the diamond operator ("<>"). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)